// @ladderlabs/buddy-sdk/dist/esm/instructions/remainingAccounts/transferRewardsNoGlobalAccount.js
import { __awaiter, __generator } from "tslib";
import { PublicKey } from "@solana/web3.js";
import { getMemberStatisticPDA, getTokenAccount } from "../../utils/pda";
import { getTreasuryPDA } from "../../utils/pda";
import { parsePublicKey } from "../../utils/parse";
import { BUDDY_DOESNT_EXIST, REFERRER_TREASURY_DOESNT_EXIST } from "../../utils/type";
export function transferRewardsNoGlobalAccount(program, mint, allowStats, memberPDA) {
    var _a;
    return __awaiter(this, void 0, void 0, function () {
        var defaultReturn, member, referrerTreasuryPDA, referrerTreasury, mintTreasuryPDA, owners, shares, _i, _b, owner, mintTreasuryATA, referrerMemberPDA, referrerMember, treasury, buddyPDA, buddyProfilePDA, buddy, referrerMemberStats;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    defaultReturn = {
                        referrerMember: PublicKey["default"],
                        referrerTreasury: PublicKey["default"],
                        referrerTreasuryRewards: PublicKey["default"],
                        referrerMemberStatistics: PublicKey["default"],
                        buddyProfile: PublicKey["default"],
                        buddy: PublicKey["default"],
                        buddyTreasury: PublicKey["default"],
                        member: PublicKey["default"],
                        referrerTokenAccount: program.programId,
                        mint: program.programId,
                        programId: program.programId
                    };
                    if (!memberPDA || memberPDA.toString() === PublicKey["default"].toString()) {
                        return [2 /*return*/, defaultReturn];
                    }
                    return [4 /*yield*/, program.account.member.fetchNullable(memberPDA)];
                case 1:
                    member = _c.sent();
                    if (!member || member.referrer.toString() === PublicKey["default"].toString()) {
                        return [2 /*return*/, defaultReturn];
                    }
                    referrerTreasuryPDA = member.referrer;
                    return [4 /*yield*/, program.account.treasury.fetchNullable(referrerTreasuryPDA)];
                case 2:
                    referrerTreasury = _c.sent();
                    if (!referrerTreasury)
                        throw REFERRER_TREASURY_DOESNT_EXIST;
                    mintTreasuryPDA = referrerTreasuryPDA;
                    if (referrerTreasury.mint.toString() !== mint.toString()) {
                        owners = [], shares = [];
                        for (_i = 0, _b = referrerTreasury.owners; _i < _b.length; _i++) {
                            owner = _b[_i];
                            owners.push(owner.ownerPda);
                            shares.push(owner.share);
                        }
                        mintTreasuryPDA = getTreasuryPDA(program, owners, shares, parsePublicKey(mint));
                    }
                    return [4 /*yield*/, getTokenAccount(mintTreasuryPDA, mint)];
                case 3:
                    mintTreasuryATA = _c.sent();
                    referrerMemberPDA = PublicKey["default"];
                    if (!allowStats) return [3 /*break*/, 5];
                    return [4 /*yield*/, program.account.member.all([
                            {
                                memcmp: {
                                    offset: 41,
                                    bytes: referrerTreasuryPDA.toBase58()
                                }
                            },
                        ])];
                case 4:
                    referrerMember = (_c.sent())[0];
                    referrerMemberPDA = (referrerMember === null || referrerMember === void 0 ? void 0 : referrerMember.publicKey) || PublicKey["default"];
                    _c.label = 5;
                case 5: return [4 /*yield*/, program.account.treasury.fetch(member.owner)];
                case 6:
                    treasury = _c.sent();
                    buddyPDA = treasury.owners[0].ownerPda;
                    buddyProfilePDA = buddyPDA;
                    return [4 /*yield*/, program.account.buddy.fetch(buddyPDA)];
                case 7:
                    buddy = _c.sent();
                    if (!buddy)
                        throw BUDDY_DOESNT_EXIST;
                    if (!((_a = buddy.buddyType) === null || _a === void 0 ? void 0 : _a.profile)) {
                        buddyProfilePDA = buddy.authority;
                    }
                    referrerMemberStats = PublicKey["default"];
                    if (referrerMemberPDA.toString() !== PublicKey["default"].toString()) {
                        referrerMemberStats = getMemberStatisticPDA(program, member.organization, referrerMemberPDA);
                    }
                    //TODO: also add the instruction to transfer folder
                    return [2 /*return*/, {
                            referrerMember: referrerMemberPDA,
                            referrerTreasury: referrerTreasuryPDA,
                            referrerTreasuryRewards: mintTreasuryPDA,
                            referrerMemberStatistics: referrerMemberStats,
                            buddyProfile: buddyProfilePDA,
                            buddy: buddyPDA,
                            buddyTreasury: member.owner,
                            member: memberPDA,
                            referrerTokenAccount: mintTreasuryATA || program.programId,
                            mint: parsePublicKey(mint) || program.programId,
                            programId: program.programId
                        }];
            }
        });
    });
}
