// @ladderlabs/buddy-sdk/dist/esm/instructions/remainingAccounts/validateReferrer.js
import { __awaiter, __generator } from "tslib";
import { PublicKey } from "@solana/web3.js";
import { getTokenAccount } from "../../utils/pda";
import { getTreasuryPDA } from "../../utils/pda";
import { parsePublicKey } from "../../utils/parse";
import { OWNER_BUDDY_DOESNT_EXIST, REFERRER_TREASURY_DOESNT_EXIST, TREASURY_DOESNT_EXIST, } from "../../utils/type";
export function validateReferrerAccounts(program, mint, allowStats, memberPDA) {
    return __awaiter(this, void 0, void 0, function () {
        var member, ownerTreasuryPDA, ownerTreasury, buddyPDA, buddy, buddyProfilePDA, referrerTreasuryPDA, referrerTreasury, mintTreasuryPDA, owners, shares, _i, _a, owner, mintTreasuryATA, referrerMemberPDA, referrerMember;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    if (!memberPDA || memberPDA.toString() === PublicKey["default"].toString()) {
                        return [2 /*return*/, {
                                programId: program.programId,
                                buddyProfile: PublicKey["default"],
                                buddy: PublicKey["default"],
                                buddyTreasury: PublicKey["default"],
                                memberPDA: PublicKey["default"],
                                referrerMember: PublicKey["default"],
                                referrerTreasury: PublicKey["default"],
                                referrerTreasuryReward: PublicKey["default"],
                                mint: program.programId,
                                referrerATA: program.programId
                            }];
                    }
                    return [4 /*yield*/, program.account.member.fetchNullable(memberPDA)];
                case 1:
                    member = _b.sent();
                    if (!member || member.referrer.toString() === PublicKey["default"].toString()) {
                        return [2 /*return*/, {
                                programId: program.programId,
                                buddyProfile: PublicKey["default"],
                                buddy: PublicKey["default"],
                                buddyTreasury: PublicKey["default"],
                                memberPDA: PublicKey["default"],
                                referrerMember: PublicKey["default"],
                                referrerTreasury: PublicKey["default"],
                                referrerTreasuryReward: PublicKey["default"],
                                mint: program.programId,
                                referrerATA: program.programId
                            }];
                    }
                    ownerTreasuryPDA = member.owner;
                    return [4 /*yield*/, program.account.treasury.fetchNullable(ownerTreasuryPDA)];
                case 2:
                    ownerTreasury = _b.sent();
                    if (!ownerTreasury)
                        throw TREASURY_DOESNT_EXIST;
                    buddyPDA = ownerTreasury.owners[0].ownerPda;
                    return [4 /*yield*/, program.account.buddy.fetchNullable(buddyPDA)];
                case 3:
                    buddy = _b.sent();
                    if (!buddy)
                        throw OWNER_BUDDY_DOESNT_EXIST;
                    buddyProfilePDA = buddyPDA;
                    if (!buddy.buddyType.profile) {
                        buddyProfilePDA = buddy.authority;
                    }
                    referrerTreasuryPDA = member.referrer;
                    return [4 /*yield*/, program.account.treasury.fetchNullable(referrerTreasuryPDA)];
                case 4:
                    referrerTreasury = _b.sent();
                    if (!referrerTreasury)
                        throw REFERRER_TREASURY_DOESNT_EXIST;
                    mintTreasuryPDA = referrerTreasuryPDA;
                    if (referrerTreasury.mint.toString() !== mint.toString()) {
                        owners = [], shares = [];
                        for (_i = 0, _a = referrerTreasury.owners; _i < _a.length; _i++) {
                            owner = _a[_i];
                            owners.push(owner.ownerPda);
                            shares.push(owner.share);
                        }
                        mintTreasuryPDA = getTreasuryPDA(program, owners, shares, parsePublicKey(mint));
                    }
                    return [4 /*yield*/, getTokenAccount(mintTreasuryPDA, mint)];
                case 5:
                    mintTreasuryATA = _b.sent();
                    referrerMemberPDA = null;
                    if (!allowStats) return [3 /*break*/, 7];
                    return [4 /*yield*/, program.account.member.all([
                            {
                                memcmp: {
                                    offset: 41,
                                    bytes: referrerTreasuryPDA.toBase58()
                                }
                            },
                        ])];
                case 6:
                    referrerMember = (_b.sent())[0];
                    referrerMemberPDA = (referrerMember === null || referrerMember === void 0 ? void 0 : referrerMember.publicKey) || null;
                    _b.label = 7;
                case 7: return [2 /*return*/, {
                        programId: program.programId,
                        buddyProfile: buddyProfilePDA,
                        buddy: buddyPDA,
                        buddyTreasury: ownerTreasuryPDA,
                        memberPDA: memberPDA,
                        referrerMember: referrerMemberPDA,
                        referrerTreasury: referrerTreasuryPDA,
                        referrerTreasuryReward: mintTreasuryPDA,
                        mint: mint.toString() === PublicKey["default"].toString()
                            ? program.programId
                            : mint,
                        referrerATA: mintTreasuryATA || program.programId
                    }];
            }
        });
    });
}
