// @ladderlabs/buddy-sdk/dist/esm/client/providers/AmbassadorProvider.js
import { __assign, __awaiter, __generator } from "tslib";
import { getAmbassadorPDA } from "../../utils/pda";
import { Ambassador } from "../../models/Ambassador";
var AmbassadorProvider = /** @class */ (function () {
    function AmbassadorProvider(program, wallet) {
        this.program = program;
        this.wallet = wallet;
    }
    AmbassadorProvider.prototype.getByOrganization = function (organization) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getByPDA(organization.account.ambassador)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    AmbassadorProvider.prototype.getByPDA = function (ambassadorPDA) {
        return __awaiter(this, void 0, void 0, function () {
            var ambassador;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.ambassador.fetchNullable(ambassadorPDA)];
                    case 1:
                        ambassador = _a.sent();
                        if (!ambassador)
                            return [2 /*return*/, ambassador];
                        return [2 /*return*/, new Ambassador(__assign(__assign({}, ambassador), { pda: ambassadorPDA }), this.program, this.wallet)];
                }
            });
        });
    };
    AmbassadorProvider.prototype.getByName = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var ambassadorPDA;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        ambassadorPDA = getAmbassadorPDA(this.program, name.toLowerCase());
                        return [4 /*yield*/, this.getByPDA(ambassadorPDA)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    return AmbassadorProvider;
}());
export { AmbassadorProvider };
