// @ladderlabs/buddy-sdk/dist/esm/client/providers/BuddyProvider.js
import { __assign, __awaiter, __generator } from "tslib";
import { Buddy } from "../../models/Buddy";
import { getBuddyPDA } from "../../utils/pda";
import { NO_WALLET_ERROR } from "../../utils/type";
var BuddyProvider = /** @class */ (function () {
    function BuddyProvider(program, wallet) {
        this.program = program;
        this.wallet = wallet;
    }
    BuddyProvider.prototype.isBuddyAvailable = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var buddyPDA;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        buddyPDA = getBuddyPDA(this.program, name.toLowerCase());
                        return [4 /*yield*/, this.program.account.buddy.fetchNullable(buddyPDA)];
                    case 1: return [2 /*return*/, !!(_a.sent())];
                }
            });
        });
    };
    BuddyProvider.prototype.getAllBuddies = function () {
        return __awaiter(this, void 0, void 0, function () {
            var buddies;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.buddy.all()];
                    case 1:
                        buddies = (_a.sent()).map(function (buddy) {
                            return new Buddy(__assign(__assign({}, buddy.account), { pda: buddy.publicKey }), _this.program, _this.wallet);
                        });
                        return [2 /*return*/, buddies];
                }
            });
        });
    };
    BuddyProvider.prototype.getByPDA = function (buddyPDA) {
        return __awaiter(this, void 0, void 0, function () {
            var buddy;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.buddy.fetchNullable(buddyPDA)];
                    case 1:
                        buddy = _a.sent();
                        if (!buddy)
                            return [2 /*return*/, null];
                        return [2 /*return*/, new Buddy(__assign(__assign({}, buddy), { pda: buddyPDA }), this.program, this.wallet)];
                }
            });
        });
    };
    BuddyProvider.prototype.getByName = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var buddyPDA;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        buddyPDA = getBuddyPDA(this.program, name.toLowerCase());
                        return [4 /*yield*/, this.getByPDA(buddyPDA)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    BuddyProvider.prototype.getAllByProfile = function (profilePDA) {
        return __awaiter(this, void 0, void 0, function () {
            var buddies;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.buddy.all([
                            { memcmp: { offset: 8, bytes: profilePDA.toBase58() } },
                        ])];
                    case 1:
                        buddies = (_a.sent()).map(function (buddy) {
                            return new Buddy(__assign(__assign({}, buddy.account), { pda: buddy.publicKey }), _this.program, _this.wallet);
                        });
                        return [2 /*return*/, buddies];
                }
            });
        });
    };
    BuddyProvider.prototype.getProfile = function (wallet) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var buddies;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!this.wallet && !wallet)
                            throw NO_WALLET_ERROR;
                        return [4 /*yield*/, this.program.account.buddy.all([
                                {
                                    memcmp: {
                                        offset: 8,
                                        bytes: wallet ? wallet.toBase58() : (_a = this.wallet) === null || _a === void 0 ? void 0 : _a.toBase58()
                                    }
                                },
                            ])];
                    case 1:
                        buddies = (_b.sent()).map(function (buddy) {
                            return new Buddy(__assign(__assign({}, buddy.account), { pda: buddy.publicKey }), _this.program, _this.wallet);
                        });
                        //always only have 1 profile
                        return [2 /*return*/, buddies[0] || null];
                }
            });
        });
    };
    return BuddyProvider;
}());
export { BuddyProvider };
