// @ladderlabs/buddy-sdk/dist/cjs/instructions/init/initTreasury.js
"use strict";
exports.__esModule = true;
exports.initTreasury = void 0;
var tslib_1 = require("tslib");
var spl_token_1 = require("@solana/spl-token");
var anchor = tslib_1.__importStar(require("@project-serum/anchor"));
var pda_1 = require("../../utils/pda");
function initTreasury(program, authority, owners, shares, mint) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var treasuryPDA, treasuryATA;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    treasuryPDA = (0, pda_1.getTreasuryPDA)(program, owners, shares, mint);
                    return [4 /*yield*/, (0, pda_1.getTokenAccount)(treasuryPDA, mint)];
                case 1:
                    treasuryATA = _a.sent();
                    return [4 /*yield*/, program.methods
                            .initBuddyTreasury(shares)
                            .accounts({
                            associatedTokenProgram: spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID,
                            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
                            systemProgram: anchor.web3.SystemProgram.programId,
                            rent: anchor.web3.SYSVAR_RENT_PUBKEY,
                            authority: authority,
                            owner1: owners[0] || null,
                            owner2: owners[1] || null,
                            owner3: owners[2] || null,
                            owner4: owners[3] || null,
                            mint: mint,
                            buddyTreasury: treasuryPDA,
                            buddyTokenAccount: treasuryATA
                        })
                            .instruction()];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    });
}
exports.initTreasury = initTreasury;
