// @ladderlabs/buddy-sdk/dist/cjs/instructions/init/initMemberStatistics.js
"use strict";
exports.__esModule = true;
exports.initMemberStatistics = void 0;
var tslib_1 = require("tslib");
var anchor = tslib_1.__importStar(require("@project-serum/anchor"));
var pda_1 = require("../../utils/pda");
var type_1 = require("../../utils/type");
function initMemberStatistics(program, authority, memberName, organization) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var memberPDA, statsPDA;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!organization)
                        throw type_1.ORGANIZATION_PDA_NOT_EXIST;
                    memberPDA = (0, pda_1.getMemberPDA)(program, organization.name, memberName);
                    statsPDA = (0, pda_1.getMemberStatisticPDA)(program, organization.name, memberPDA);
                    return [4 /*yield*/, program.methods.initMemberStatistics()
                            .accounts({
                            systemProgram: anchor.web3.SystemProgram.programId,
                            rent: anchor.web3.SYSVAR_RENT_PUBKEY,
                            authority: authority,
                            organization: organization.pda,
                            member: memberPDA,
                            memberStatistics: statsPDA
                        })
                            .instruction()];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    });
}
exports.initMemberStatistics = initMemberStatistics;
