// @ladderlabs/buddy-sdk/dist/cjs/instructions/init/initMasterTreasury.js
"use strict";
exports.__esModule = true;
exports.initMasterTreasury = void 0;
var tslib_1 = require("tslib");
var spl_token_1 = require("@solana/spl-token");
var anchor = tslib_1.__importStar(require("@project-serum/anchor"));
var pda_1 = require("../../utils/pda");
var type_1 = require("../../utils/type");
function initMasterTreasury(program, authority, mint) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var masterOrg, treasuryPDA, treasuryATA;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, program.account.masterOrganization.all()];
                case 1:
                    masterOrg = (_a.sent())[0];
                    treasuryPDA = anchor.web3.PublicKey.findProgramAddressSync([Buffer.from(type_1.TREASURY_NAME), Buffer.from(type_1.MASTER_NAME), mint.toBuffer()], program.programId)[0];
                    return [4 /*yield*/, (0, pda_1.getTokenAccount)(treasuryPDA, mint)];
                case 2:
                    treasuryATA = _a.sent();
                    return [4 /*yield*/, program.methods
                            .initMasterOrganizationTreasury()
                            .accounts({
                            associatedTokenProgram: spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID,
                            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
                            systemProgram: anchor.web3.SystemProgram.programId,
                            rent: anchor.web3.SYSVAR_RENT_PUBKEY,
                            authority: authority,
                            masterOrganization: masterOrg.publicKey,
                            mint: mint,
                            masterOrganizationTreasury: treasuryPDA,
                            masterOrganizationTokenAccount: treasuryATA
                        })
                            .instruction()];
                case 3: return [2 /*return*/, _a.sent()];
            }
        });
    });
}
exports.initMasterTreasury = initMasterTreasury;
