// @ladderlabs/buddy-sdk/dist/cjs/instructions/init/initGlobalBuddy.js
"use strict";
exports.__esModule = true;
exports.initGlobalBuddy = void 0;
var tslib_1 = require("tslib");
var spl_token_1 = require("@solana/spl-token");
var anchor = tslib_1.__importStar(require("@project-serum/anchor"));
var pda_1 = require("../../utils/pda");
var initTreasury_1 = require("./initTreasury");
var accounts_1 = require("../../utils/accounts");
var type_1 = require("../../utils/type");
function initGlobalBuddy(program, authority, globalBuddyName, buddyProfile, usdcMint) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var instructions, buddyPDA, buddyNamePDA, walletATA, referrerTreasury, owners, shares, _i, _a, owner, referrerTreasuryUSDCPDA, referrerTreasuryUSDCATA, referrerTreasuryUSDC, _b, _c, usdcATA, _d, _e;
        return tslib_1.__generator(this, function (_f) {
            switch (_f.label) {
                case 0:
                    instructions = [];
                    buddyPDA = (0, pda_1.getBuddyPDA)(program, globalBuddyName);
                    buddyNamePDA = (0, pda_1.getBuddyNamePDA)(program, globalBuddyName);
                    return [4 /*yield*/, (0, pda_1.getTokenAccount)(authority, usdcMint, false)];
                case 1:
                    walletATA = _f.sent();
                    return [4 /*yield*/, program.account.treasury.fetchNullable(buddyProfile.referrerTreasuryPda)];
                case 2:
                    referrerTreasury = _f.sent();
                    if (!referrerTreasury)
                        throw type_1.REFERRER_TREASURY_DOESNT_EXIST;
                    owners = [], shares = [];
                    for (_i = 0, _a = referrerTreasury.owners; _i < _a.length; _i++) {
                        owner = _a[_i];
                        owners.push(owner.ownerPda);
                        shares.push(owner.share);
                    }
                    referrerTreasuryUSDCPDA = (0, pda_1.getTreasuryPDA)(program, owners, shares, usdcMint);
                    return [4 /*yield*/, (0, pda_1.getTokenAccount)(referrerTreasuryUSDCPDA, usdcMint)];
                case 3:
                    referrerTreasuryUSDCATA = _f.sent();
                    if (!(referrerTreasury.mint.toString() !== usdcMint.toString())) return [3 /*break*/, 6];
                    return [4 /*yield*/, program.account.treasury.fetchNullable(referrerTreasuryUSDCPDA)];
                case 4:
                    referrerTreasuryUSDC = _f.sent();
                    if (!!referrerTreasuryUSDC) return [3 /*break*/, 6];
                    _c = (_b = instructions).push;
                    return [4 /*yield*/, (0, initTreasury_1.initTreasury)(program, authority, owners, shares, usdcMint)];
                case 5:
                    _c.apply(_b, [_f.sent()]);
                    _f.label = 6;
                case 6: return [4 /*yield*/, (0, accounts_1.getBuddylinkBuddyAccounts)(program, "", usdcMint)];
                case 7:
                    usdcATA = (_f.sent()).ata;
                    _e = (_d = instructions).push;
                    return [4 /*yield*/, program.methods
                            .initBuddyPaid(globalBuddyName)
                            .accounts({
                            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
                            systemProgram: anchor.web3.SystemProgram.programId,
                            rent: anchor.web3.SYSVAR_RENT_PUBKEY,
                            authority: authority,
                            buddyProfile: buddyProfile.pda,
                            referrerSharedTreasury: buddyProfile.referrerTreasuryPda,
                            buddy: buddyPDA,
                            buddyName: buddyNamePDA,
                            usdcMint: usdcMint,
                            usdcTokenAccount: walletATA,
                            buddylinkUsdcTokenAccount: usdcATA,
                            referrerUsdcSharedTreasury: referrerTreasuryUSDCPDA,
                            referrerUsdcTokenAccount: referrerTreasuryUSDCATA
                        })
                            .instruction()];
                case 8:
                    _e.apply(_d, [_f.sent()]);
                    return [2 /*return*/, instructions];
            }
        });
    });
}
exports.initGlobalBuddy = initGlobalBuddy;
