// @ladderlabs/buddy-sdk/dist/esm/utils/type.js
import { PublicKey } from "@solana/web3.js";
import { PROGRAM_ID as MP_METADATA_PROGRAM_ID } from "@metaplex-foundation/mpl-token-metadata";
export var METADATA_PROGRAM_ID = MP_METADATA_PROGRAM_ID;
export var DECIMALS = 1000000000;
export var FORCED_REFERRER_BPS = 1000;
export var MINT_SOLANA_NAME = "solana_mint_";
export var MASTER_NAME = "master_";
export var ORGANIZATION_NAME = "organization_";
export var BUDDY_NAME = "buddy_";
export var MEMBER_NAME = "member_";
export var AMBASSADOR_NAME = "ambassador_";
export var TREASURY_NAME = "treasury_";
export var BUDDY_LINK_BUDDY_NAME = "buddylink_";
export var ORGANIZATION_NAME_CHECK = "name_organization_";
export var BUDDY_NAME_CHECK = "name_buddy_";
export var MEMBER_NAME_CHECK = "name_member_";
export var PAID_BUDDY_CHECK = "paid_";
export var MAINNET_PROGRAM_ID = "BUDDYtQp7Di1xfojiCSVDksiYLQx511DPdj2nbtG9Yu5";
export var USDC_MINT = new PublicKey("EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v");
export var PROGRAM_USDC_ATA = new PublicKey("7HXsteWfPHDq8PJ5ghYvqq3MpFUQJtx3kQLvLNc9bdaP");
export var COLLECTION_MINT = new PublicKey("7HXsteWfPHDq8PJ5ghYvqq3MpFUQJtx3kQLvLNc9bdaP"); // TODO: change to correct one
// Errors
export var NO_WALLET_ERROR = "No wallet passed.";
export var MEMBER_ERROR = "Member \"{}\" doesn't exist";
export var ORGANIZATION_NOT_EXIST = "Organization \"{}\" doesn't exist";
export var ORGANIZATION_PDA_NOT_EXIST = "The key given is not a valid organization.";
export var MEMBER_NOT_EXIST = "The key given is not a valid member.";
export var ONLY_PROFILE_BUY = "Only the buddy profile can buy a global buddy.";
export var TREASURY_OWNER_MISMATCH = "The buddy is not an owner of this treasury.";
export var BUDDY_NO_REFERRER = "The buddy doesn't have a referrer to send funds to.";
export var BUDDY_DOESNT_EXIST = "The key given is not a valid buddy.";
export var OWNER_BUDDY_DOESNT_EXIST = "CRITICAL: The key given is not a valid owner buddy.";
export var TREASURY_DOESNT_EXIST = "The key given is not a valid treasury.";
export var REFERRER_TREASURY_DOESNT_EXIST = "CRITICAL: The key given is not a valid referrer treasury.";
