// @ladderlabs/buddy-sdk/dist/esm/client/builders/Accounts.js
import { __awaiter, __generator } from "tslib";
import { PublicKey } from "@solana/web3.js";
import { validateReferrerAccounts } from "../../instructions/remainingAccounts/validateReferrer";
import { transferRewardsAccount } from "../../instructions/remainingAccounts/transferRewardsAccounts";
import { parsePublicKey } from "../../utils/parse";
import { transferRewardsNoGlobalAccount } from "../../instructions/remainingAccounts/transferRewardsNoGlobalAccount";
import { transferRewardsOptimisticAccount } from "../../instructions/remainingAccounts/transferRewardsOptimisticAccount";
var Accounts = /** @class */ (function () {
    // private wallet?: PublicKey;
    function Accounts(program) {
        this.program = program;
        // this.wallet = wallet;
    }
    Accounts.prototype.validateReferrerAccounts = function (mint, memberPDA, allowStats) {
        if (allowStats === void 0) { allowStats = true; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, validateReferrerAccounts(this.program, mint, allowStats, memberPDA)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Accounts.prototype.transferRewardsAccount = function (memberPDA, mint, allowStats) {
        if (allowStats === void 0) { allowStats = true; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, transferRewardsAccount(this.program, parsePublicKey(mint) ? mint : PublicKey["default"], allowStats, memberPDA)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Accounts.prototype.transferRewardsNoGlobalAccount = function (memberPDA, mint, allowStats) {
        if (allowStats === void 0) { allowStats = true; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, transferRewardsNoGlobalAccount(this.program, parsePublicKey(mint) ? mint : PublicKey["default"], allowStats, memberPDA)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Accounts.prototype.transferRewardsOptimisticAccount = function (profileName, // can only be profile
    memberName, organizationName, mint, referrerName) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, transferRewardsOptimisticAccount(this.program, profileName, memberName, organizationName, mint, referrerName)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    return Accounts;
}());
export { Accounts };
