// @ladderlabs/buddy-sdk/dist/cjs/models/Buddy.js
"use strict";
exports.__esModule = true;
exports.Buddy = void 0;
var tslib_1 = require("tslib");
var type_1 = require("../utils/type");
var initGlobalBuddy_1 = require("../instructions/init/initGlobalBuddy");
var Buddy = /** @class */ (function () {
    function Buddy(buddy, program, wallet) {
        this.account = buddy;
        this.program = program;
        this.wallet = wallet;
    }
    Buddy.prototype.refresh = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, _b;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _a = this;
                        _b = [{ pda: this.account.pda }];
                        return [4 /*yield*/, this.program.account.buddy.fetch(this.account.pda)];
                    case 1:
                        _a.account = tslib_1.__assign.apply(void 0, _b.concat([(_c.sent())]));
                        return [2 /*return*/, this];
                }
            });
        });
    };
    Buddy.prototype.buyGlobalBuddy = function (buddyName) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.wallet)
                            throw type_1.NO_WALLET_ERROR;
                        if (!this.account.buddyType.profile)
                            throw type_1.ONLY_PROFILE_BUY;
                        return [4 /*yield*/, (0, initGlobalBuddy_1.initGlobalBuddy)(this.program, this.wallet, buddyName, this.account, this.usdcMint ? this.usdcMint : type_1.USDC_MINT)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Buddy.prototype.inject = function (usdcMint) {
        this.usdcMint = usdcMint;
        return this;
    };
    return Buddy;
}());
exports.Buddy = Buddy;
