// @ladderlabs/buddy-sdk/dist/cjs/utils/type.js
"use strict";
exports.__esModule = true;
exports.REFERRER_TREASURY_DOESNT_EXIST = exports.TREASURY_DOESNT_EXIST = exports.OWNER_BUDDY_DOESNT_EXIST = exports.BUDDY_DOESNT_EXIST = exports.BUDDY_NO_REFERRER = exports.TREASURY_OWNER_MISMATCH = exports.ONLY_PROFILE_BUY = exports.MEMBER_NOT_EXIST = exports.ORGANIZATION_PDA_NOT_EXIST = exports.ORGANIZATION_NOT_EXIST = exports.MEMBER_ERROR = exports.NO_WALLET_ERROR = exports.COLLECTION_MINT = exports.PROGRAM_USDC_ATA = exports.USDC_MINT = exports.MAINNET_PROGRAM_ID = exports.PAID_BUDDY_CHECK = exports.MEMBER_NAME_CHECK = exports.BUDDY_NAME_CHECK = exports.ORGANIZATION_NAME_CHECK = exports.BUDDY_LINK_BUDDY_NAME = exports.TREASURY_NAME = exports.AMBASSADOR_NAME = exports.MEMBER_NAME = exports.BUDDY_NAME = exports.ORGANIZATION_NAME = exports.MASTER_NAME = exports.MINT_SOLANA_NAME = exports.FORCED_REFERRER_BPS = exports.DECIMALS = exports.METADATA_PROGRAM_ID = void 0;
var web3_js_1 = require("@solana/web3.js");
var mpl_token_metadata_1 = require("@metaplex-foundation/mpl-token-metadata");
exports.METADATA_PROGRAM_ID = mpl_token_metadata_1.PROGRAM_ID;
exports.DECIMALS = 1000000000;
exports.FORCED_REFERRER_BPS = 1000;
exports.MINT_SOLANA_NAME = "solana_mint_";
exports.MASTER_NAME = "master_";
exports.ORGANIZATION_NAME = "organization_";
exports.BUDDY_NAME = "buddy_";
exports.MEMBER_NAME = "member_";
exports.AMBASSADOR_NAME = "ambassador_";
exports.TREASURY_NAME = "treasury_";
exports.BUDDY_LINK_BUDDY_NAME = "buddylink_";
exports.ORGANIZATION_NAME_CHECK = "name_organization_";
exports.BUDDY_NAME_CHECK = "name_buddy_";
exports.MEMBER_NAME_CHECK = "name_member_";
exports.PAID_BUDDY_CHECK = "paid_";
exports.MAINNET_PROGRAM_ID = "BUDDYtQp7Di1xfojiCSVDksiYLQx511DPdj2nbtG9Yu5";
exports.USDC_MINT = new web3_js_1.PublicKey("EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v");
exports.PROGRAM_USDC_ATA = new web3_js_1.PublicKey("7HXsteWfPHDq8PJ5ghYvqq3MpFUQJtx3kQLvLNc9bdaP");
exports.COLLECTION_MINT = new web3_js_1.PublicKey("7HXsteWfPHDq8PJ5ghYvqq3MpFUQJtx3kQLvLNc9bdaP"); // TODO: change to correct one
// Errors
exports.NO_WALLET_ERROR = "No wallet passed.";
exports.MEMBER_ERROR = "Member \"{}\" doesn't exist";
exports.ORGANIZATION_NOT_EXIST = "Organization \"{}\" doesn't exist";
exports.ORGANIZATION_PDA_NOT_EXIST = "The key given is not a valid organization.";
exports.MEMBER_NOT_EXIST = "The key given is not a valid member.";
exports.ONLY_PROFILE_BUY = "Only the buddy profile can buy a global buddy.";
exports.TREASURY_OWNER_MISMATCH = "The buddy is not an owner of this treasury.";
exports.BUDDY_NO_REFERRER = "The buddy doesn't have a referrer to send funds to.";
exports.BUDDY_DOESNT_EXIST = "The key given is not a valid buddy.";
exports.OWNER_BUDDY_DOESNT_EXIST = "CRITICAL: The key given is not a valid owner buddy.";
exports.TREASURY_DOESNT_EXIST = "The key given is not a valid treasury.";
exports.REFERRER_TREASURY_DOESNT_EXIST = "CRITICAL: The key given is not a valid referrer treasury.";
