// @ladderlabs/buddy-sdk/dist/cjs/client/builders/Accounts.js
"use strict";
exports.__esModule = true;
exports.Accounts = void 0;
var tslib_1 = require("tslib");
var web3_js_1 = require("@solana/web3.js");
var validateReferrer_1 = require("../../instructions/remainingAccounts/validateReferrer");
var transferRewardsAccounts_1 = require("../../instructions/remainingAccounts/transferRewardsAccounts");
var parse_1 = require("../../utils/parse");
var transferRewardsNoGlobalAccount_1 = require("../../instructions/remainingAccounts/transferRewardsNoGlobalAccount");
var transferRewardsOptimisticAccount_1 = require("../../instructions/remainingAccounts/transferRewardsOptimisticAccount");
var Accounts = /** @class */ (function () {
    // private wallet?: PublicKey;
    function Accounts(program) {
        this.program = program;
        // this.wallet = wallet;
    }
    Accounts.prototype.validateReferrerAccounts = function (mint, memberPDA, allowStats) {
        if (allowStats === void 0) { allowStats = true; }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, validateReferrer_1.validateReferrerAccounts)(this.program, mint, allowStats, memberPDA)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Accounts.prototype.transferRewardsAccount = function (memberPDA, mint, allowStats) {
        if (allowStats === void 0) { allowStats = true; }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, transferRewardsAccounts_1.transferRewardsAccount)(this.program, (0, parse_1.parsePublicKey)(mint) ? mint : web3_js_1.PublicKey["default"], allowStats, memberPDA)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Accounts.prototype.transferRewardsNoGlobalAccount = function (memberPDA, mint, allowStats) {
        if (allowStats === void 0) { allowStats = true; }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, transferRewardsNoGlobalAccount_1.transferRewardsNoGlobalAccount)(this.program, (0, parse_1.parsePublicKey)(mint) ? mint : web3_js_1.PublicKey["default"], allowStats, memberPDA)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Accounts.prototype.transferRewardsOptimisticAccount = function (profileName, // can only be profile
    memberName, organizationName, mint, referrerName) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, transferRewardsOptimisticAccount_1.transferRewardsOptimisticAccount)(this.program, profileName, memberName, organizationName, mint, referrerName)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    return Accounts;
}());
exports.Accounts = Accounts;
