// @ladderlabs/buddy-sdk/dist/esm/instructions/init/initMemberStatistics.js
import { __awaiter, __generator } from "tslib";
import * as anchor from "@project-serum/anchor";
import { getMemberPDA, getMemberStatisticPDA, } from "../../utils/pda";
import { ORGANIZATION_PDA_NOT_EXIST } from "../../utils/type";
export function initMemberStatistics(program, authority, memberName, organization) {
    return __awaiter(this, void 0, void 0, function () {
        var memberPDA, statsPDA;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!organization)
                        throw ORGANIZATION_PDA_NOT_EXIST;
                    memberPDA = getMemberPDA(program, organization.name, memberName);
                    statsPDA = getMemberStatisticPDA(program, organization.name, memberPDA);
                    return [4 /*yield*/, program.methods.initMemberStatistics()
                            .accounts({
                            systemProgram: anchor.web3.SystemProgram.programId,
                            rent: anchor.web3.SYSVAR_RENT_PUBKEY,
                            authority: authority,
                            organization: organization.pda,
                            member: memberPDA,
                            memberStatistics: statsPDA
                        })
                            .instruction()];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    });
}
