// @ladderlabs/buddy-sdk/dist/cjs/instructions/remainingAccounts/transferRewardsNoGlobalAccount.js
"use strict";
exports.__esModule = true;
exports.transferRewardsNoGlobalAccount = void 0;
var tslib_1 = require("tslib");
var web3_js_1 = require("@solana/web3.js");
var pda_1 = require("../../utils/pda");
var pda_2 = require("../../utils/pda");
var parse_1 = require("../../utils/parse");
var type_1 = require("../../utils/type");
function transferRewardsNoGlobalAccount(program, mint, allowStats, memberPDA) {
    var _a;
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var defaultReturn, member, referrerTreasuryPDA, referrerTreasury, mintTreasuryPDA, owners, shares, _i, _b, owner, mintTreasuryATA, referrerMemberPDA, referrerMember, treasury, buddyPDA, buddyProfilePDA, buddy, referrerMemberStats;
        return tslib_1.__generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    defaultReturn = {
                        referrerMember: web3_js_1.PublicKey["default"],
                        referrerTreasury: web3_js_1.PublicKey["default"],
                        referrerTreasuryRewards: web3_js_1.PublicKey["default"],
                        referrerMemberStatistics: web3_js_1.PublicKey["default"],
                        buddyProfile: web3_js_1.PublicKey["default"],
                        buddy: web3_js_1.PublicKey["default"],
                        buddyTreasury: web3_js_1.PublicKey["default"],
                        member: web3_js_1.PublicKey["default"],
                        referrerTokenAccount: program.programId,
                        mint: program.programId,
                        programId: program.programId
                    };
                    if (!memberPDA || memberPDA.toString() === web3_js_1.PublicKey["default"].toString()) {
                        return [2 /*return*/, defaultReturn];
                    }
                    return [4 /*yield*/, program.account.member.fetchNullable(memberPDA)];
                case 1:
                    member = _c.sent();
                    if (!member || member.referrer.toString() === web3_js_1.PublicKey["default"].toString()) {
                        return [2 /*return*/, defaultReturn];
                    }
                    referrerTreasuryPDA = member.referrer;
                    return [4 /*yield*/, program.account.treasury.fetchNullable(referrerTreasuryPDA)];
                case 2:
                    referrerTreasury = _c.sent();
                    if (!referrerTreasury)
                        throw type_1.REFERRER_TREASURY_DOESNT_EXIST;
                    mintTreasuryPDA = referrerTreasuryPDA;
                    if (referrerTreasury.mint.toString() !== mint.toString()) {
                        owners = [], shares = [];
                        for (_i = 0, _b = referrerTreasury.owners; _i < _b.length; _i++) {
                            owner = _b[_i];
                            owners.push(owner.ownerPda);
                            shares.push(owner.share);
                        }
                        mintTreasuryPDA = (0, pda_2.getTreasuryPDA)(program, owners, shares, (0, parse_1.parsePublicKey)(mint));
                    }
                    return [4 /*yield*/, (0, pda_1.getTokenAccount)(mintTreasuryPDA, mint)];
                case 3:
                    mintTreasuryATA = _c.sent();
                    referrerMemberPDA = web3_js_1.PublicKey["default"];
                    if (!allowStats) return [3 /*break*/, 5];
                    return [4 /*yield*/, program.account.member.all([
                            {
                                memcmp: {
                                    offset: 41,
                                    bytes: referrerTreasuryPDA.toBase58()
                                }
                            },
                        ])];
                case 4:
                    referrerMember = (_c.sent())[0];
                    referrerMemberPDA = (referrerMember === null || referrerMember === void 0 ? void 0 : referrerMember.publicKey) || web3_js_1.PublicKey["default"];
                    _c.label = 5;
                case 5: return [4 /*yield*/, program.account.treasury.fetch(member.owner)];
                case 6:
                    treasury = _c.sent();
                    buddyPDA = treasury.owners[0].ownerPda;
                    buddyProfilePDA = buddyPDA;
                    return [4 /*yield*/, program.account.buddy.fetch(buddyPDA)];
                case 7:
                    buddy = _c.sent();
                    if (!buddy)
                        throw type_1.BUDDY_DOESNT_EXIST;
                    if (!((_a = buddy.buddyType) === null || _a === void 0 ? void 0 : _a.profile)) {
                        buddyProfilePDA = buddy.authority;
                    }
                    referrerMemberStats = web3_js_1.PublicKey["default"];
                    if (referrerMemberPDA.toString() !== web3_js_1.PublicKey["default"].toString()) {
                        referrerMemberStats = (0, pda_1.getMemberStatisticPDA)(program, member.organization, referrerMemberPDA);
                    }
                    //TODO: also add the instruction to transfer folder
                    return [2 /*return*/, {
                            referrerMember: referrerMemberPDA,
                            referrerTreasury: referrerTreasuryPDA,
                            referrerTreasuryRewards: mintTreasuryPDA,
                            referrerMemberStatistics: referrerMemberStats,
                            buddyProfile: buddyProfilePDA,
                            buddy: buddyPDA,
                            buddyTreasury: member.owner,
                            member: memberPDA,
                            referrerTokenAccount: mintTreasuryATA || program.programId,
                            mint: (0, parse_1.parsePublicKey)(mint) || program.programId,
                            programId: program.programId
                        }];
            }
        });
    });
}
exports.transferRewardsNoGlobalAccount = transferRewardsNoGlobalAccount;
