// @ladderlabs/buddy-sdk/dist/cjs/instructions/remainingAccounts/transferRewardsOptimisticAccount.js
"use strict";
exports.__esModule = true;
exports.transferRewardsOptimisticAccount = void 0;
var tslib_1 = require("tslib");
var web3_js_1 = require("@solana/web3.js");
var parse_1 = require("../../utils/parse");
var pda_1 = require("../../utils/pda");
var Client_1 = require("../../client/Client");
function transferRewardsOptimisticAccount(program, profileName, // can only be profile
memberName, organizationName, mint, referrerHash) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var defaultReturn, organizationPDA, organization, referrerMemberPDA, referrerMember, referrerTreasuryPDA, _a, buddyNames, shares, buddyPDAs, referrerBuddyPDA, referrerTreasury, referrerTreasuryRewardsPDA, shares, owners, _i, _b, owner, profilePDA, memberPDA, treasuryPDA, referrerTokenAccount, _c, referrerMemberStats;
        return tslib_1.__generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    defaultReturn = {
                        referrerMember: web3_js_1.PublicKey["default"],
                        referrerTreasury: web3_js_1.PublicKey["default"],
                        referrerTreasuryRewards: web3_js_1.PublicKey["default"],
                        referrerMemberStatistics: web3_js_1.PublicKey["default"],
                        buddyProfile: web3_js_1.PublicKey["default"],
                        buddy: web3_js_1.PublicKey["default"],
                        buddyTreasury: web3_js_1.PublicKey["default"],
                        member: web3_js_1.PublicKey["default"],
                        referrerTokenAccount: program.programId,
                        mint: program.programId,
                        programId: program.programId
                    };
                    if (!referrerHash)
                        return [2 /*return*/, defaultReturn];
                    organizationPDA = (0, pda_1.getOrganizationPDA)(program, organizationName);
                    return [4 /*yield*/, program.account.organization.fetch(organizationPDA)];
                case 1:
                    organization = _d.sent();
                    referrerMemberPDA = web3_js_1.PublicKey["default"];
                    referrerMember = null;
                    referrerTreasuryPDA = web3_js_1.PublicKey["default"];
                    if (!referrerHash.endsWith('$')) return [3 /*break*/, 2];
                    _a = Client_1.Client.decryptHash(referrerHash), buddyNames = _a[0], shares = _a[1];
                    buddyPDAs = buddyNames.map(function (buddyName) { return (0, pda_1.getBuddyPDA)(program, buddyName); });
                    referrerTreasuryPDA = (0, pda_1.getTreasuryPDA)(program, buddyPDAs, shares, organization.mainTokenMint);
                    return [3 /*break*/, 6];
                case 2:
                    referrerMemberPDA = (0, pda_1.getMemberPDA)(program, organizationName, referrerHash);
                    return [4 /*yield*/, program.account.member.fetchNullable(referrerMemberPDA)];
                case 3:
                    referrerMember = _d.sent();
                    if (!referrerMember) return [3 /*break*/, 4];
                    referrerTreasuryPDA = referrerMember.owner;
                    return [3 /*break*/, 6];
                case 4:
                    referrerBuddyPDA = (0, pda_1.getBuddyPDA)(program, referrerHash);
                    referrerTreasuryPDA = (0, pda_1.getTreasuryPDA)(program, [referrerBuddyPDA], [10000], organization.mainTokenMint);
                    return [4 /*yield*/, program.account.member.all([
                            { memcmp: { offset: 41, bytes: referrerTreasuryPDA.toBase58() } },
                        ])];
                case 5:
                    referrerMember = (_d.sent())[0];
                    referrerMemberPDA = (referrerMember === null || referrerMember === void 0 ? void 0 : referrerMember.publicKey) || web3_js_1.PublicKey["default"];
                    _d.label = 6;
                case 6: return [4 /*yield*/, program.account.treasury.fetchNullable(referrerTreasuryPDA)];
                case 7:
                    referrerTreasury = _d.sent();
                    if (!referrerTreasury)
                        return [2 /*return*/, defaultReturn];
                    referrerTreasuryRewardsPDA = referrerTreasuryPDA;
                    if (referrerTreasury.mint.toString() !== mint.toString()) {
                        shares = [], owners = [];
                        for (_i = 0, _b = referrerTreasury.owners; _i < _b.length; _i++) {
                            owner = _b[_i];
                            shares.push(owner.share);
                            owners.push(owner.ownerPda);
                        }
                        referrerTreasuryRewardsPDA = (0, pda_1.getTreasuryPDA)(program, owners, shares, (0, parse_1.parsePublicKey)(mint));
                    }
                    profilePDA = (0, pda_1.getBuddyPDA)(program, profileName);
                    memberPDA = (0, pda_1.getMemberPDA)(program, organizationName, memberName);
                    treasuryPDA = (0, pda_1.getTreasuryPDA)(program, [profilePDA], [10000], organization.mainTokenMint);
                    if (!((0, parse_1.parsePublicKey)(mint) && referrerTreasuryRewardsPDA)) return [3 /*break*/, 9];
                    return [4 /*yield*/, (0, pda_1.getTokenAccount)(referrerTreasuryRewardsPDA, mint)];
                case 8:
                    _c = _d.sent();
                    return [3 /*break*/, 10];
                case 9:
                    _c = program.programId;
                    _d.label = 10;
                case 10:
                    referrerTokenAccount = _c;
                    referrerMemberStats = web3_js_1.PublicKey["default"];
                    if (referrerMemberPDA.toString() !== web3_js_1.PublicKey["default"].toString()) {
                        referrerMemberStats = (0, pda_1.getMemberStatisticPDA)(program, organizationName, referrerMemberPDA);
                    }
                    return [2 /*return*/, {
                            referrerMember: referrerMemberPDA,
                            referrerTreasury: referrerTreasuryPDA,
                            referrerTreasuryRewards: referrerTreasuryRewardsPDA,
                            referrerMemberStatistics: referrerMemberStats,
                            buddyProfile: profilePDA,
                            buddy: profilePDA,
                            buddyTreasury: treasuryPDA,
                            member: memberPDA,
                            referrerTokenAccount: referrerTokenAccount,
                            mint: (0, parse_1.parsePublicKey)(mint) || program.programId,
                            programId: program.programId
                        }];
            }
        });
    });
}
exports.transferRewardsOptimisticAccount = transferRewardsOptimisticAccount;
