// @ladderlabs/buddy-sdk/dist/cjs/instructions/transfers/transferRewardsAccounts.js
"use strict";
exports.__esModule = true;
exports.transferRewardsAccount = void 0;
var tslib_1 = require("tslib");
var web3_js_1 = require("@solana/web3.js");
var pda_1 = require("../../utils/pda");
var pda_2 = require("../../utils/pda");
var parse_1 = require("../../utils/parse");
var type_1 = require("../../utils/type");
function transferRewardsAccount(program, mint, allowStats, memberPDA) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var member, referrerTreasuryPDA, referrerTreasury, mintTreasuryPDA, owners, shares, _i, _a, owner, mintTreasuryATA, mintGlobalReferrerTreasuryATA, mintGlobalReferrerTreasuryPDA, globalReferrerTreasuryPDA, globalReferrerTreasury, owners, shares, _b, _c, owner, referrerMemberPDA, referrerMember;
        return tslib_1.__generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    if (!memberPDA || memberPDA.toString() === web3_js_1.PublicKey["default"].toString()) {
                        return [2 /*return*/, {
                                referrerAccount: web3_js_1.PublicKey["default"],
                                referrerAccountRewards: web3_js_1.PublicKey["default"],
                                referreeAccount: web3_js_1.PublicKey["default"],
                                globalReferrerTreasury: web3_js_1.PublicKey["default"],
                                globalReferrerTokenAccount: program.programId,
                                mint: program.programId,
                                referrerTokenAccount: program.programId
                            }];
                    }
                    return [4 /*yield*/, program.account.member.fetchNullable(memberPDA)];
                case 1:
                    member = _d.sent();
                    if (!member || member.referrer.toString() === web3_js_1.PublicKey["default"].toString()) {
                        return [2 /*return*/, {
                                referrerAccount: web3_js_1.PublicKey["default"],
                                referrerAccountRewards: web3_js_1.PublicKey["default"],
                                referreeAccount: web3_js_1.PublicKey["default"],
                                globalReferrerTreasury: web3_js_1.PublicKey["default"],
                                globalReferrerTokenAccount: program.programId,
                                mint: program.programId,
                                referrerTokenAccount: program.programId
                            }];
                    }
                    referrerTreasuryPDA = member.referrer;
                    return [4 /*yield*/, program.account.treasury.fetchNullable(referrerTreasuryPDA)];
                case 2:
                    referrerTreasury = _d.sent();
                    if (!referrerTreasury)
                        throw type_1.REFERRER_TREASURY_DOESNT_EXIST;
                    mintTreasuryPDA = referrerTreasuryPDA;
                    if (referrerTreasury.mint.toString() !== mint.toString()) {
                        owners = [], shares = [];
                        for (_i = 0, _a = referrerTreasury.owners; _i < _a.length; _i++) {
                            owner = _a[_i];
                            owners.push(owner.ownerPda);
                            shares.push(owner.share);
                        }
                        mintTreasuryPDA = (0, pda_2.getTreasuryPDA)(program, owners, shares, (0, parse_1.parsePublicKey)(mint));
                    }
                    return [4 /*yield*/, (0, pda_1.getTokenAccount)(mintTreasuryPDA, mint)];
                case 3:
                    mintTreasuryATA = _d.sent();
                    mintGlobalReferrerTreasuryATA = null;
                    mintGlobalReferrerTreasuryPDA = null;
                    if (!(JSON.stringify(member.globalReferrerOfBuddy.referrerTreasuryOwnerHash) !==
                        JSON.stringify(new Array(32).fill(0)))) return [3 /*break*/, 6];
                    globalReferrerTreasuryPDA = (0, pda_1.getTreasuryPDAByHash)(program, Buffer.from(member.globalReferrerOfBuddy.referrerTreasuryOwnerHash), (0, parse_1.parsePublicKey)(mint));
                    return [4 /*yield*/, program.account.treasury.fetchNullable(globalReferrerTreasuryPDA)];
                case 4:
                    globalReferrerTreasury = _d.sent();
                    if (!globalReferrerTreasury)
                        throw type_1.REFERRER_TREASURY_DOESNT_EXIST;
                    owners = [], shares = [];
                    for (_b = 0, _c = globalReferrerTreasury.owners; _b < _c.length; _b++) {
                        owner = _c[_b];
                        owners.push(owner.ownerPda);
                        shares.push(owner.share);
                    }
                    mintGlobalReferrerTreasuryPDA = (0, pda_2.getTreasuryPDA)(program, owners, shares, (0, parse_1.parsePublicKey)(mint));
                    return [4 /*yield*/, (0, pda_1.getTokenAccount)(mintGlobalReferrerTreasuryPDA, mint)];
                case 5:
                    mintGlobalReferrerTreasuryATA = _d.sent();
                    _d.label = 6;
                case 6:
                    referrerMemberPDA = null;
                    if (!allowStats) return [3 /*break*/, 8];
                    return [4 /*yield*/, program.account.member.all([
                            {
                                memcmp: {
                                    offset: 41,
                                    bytes: referrerTreasuryPDA.toBase58()
                                }
                            },
                        ])];
                case 7:
                    referrerMember = (_d.sent())[0];
                    referrerMemberPDA = (referrerMember === null || referrerMember === void 0 ? void 0 : referrerMember.publicKey) || null;
                    _d.label = 8;
                case 8: return [2 /*return*/, {
                        referrerMemberAccount: referrerMemberPDA,
                        referrerAccount: referrerTreasuryPDA,
                        referrerAccountRewards: mintTreasuryPDA,
                        referreeAccount: memberPDA,
                        globalReferrerTreasury: mintGlobalReferrerTreasuryPDA || program.programId,
                        globalReferrerTokenAccount: mintGlobalReferrerTreasuryATA || program.programId,
                        mint: (0, parse_1.parsePublicKey)(mint) || program.programId,
                        referrerTokenAccount: mintTreasuryATA || program.programId
                    }];
            }
        });
    });
}
exports.transferRewardsAccount = transferRewardsAccount;
