// @ladderlabs/buddy-sdk/dist/cjs/instructions/transfers/transferUp.js
"use strict";
exports.__esModule = true;
exports.transferUp = void 0;
var tslib_1 = require("tslib");
var spl_token_1 = require("@solana/spl-token");
var anchor = tslib_1.__importStar(require("@project-serum/anchor"));
var pda_1 = require("../../utils/pda");
var pda_2 = require("../../utils/pda");
var type_1 = require("../../utils/type");
var parse_1 = require("../../utils/parse");
function transferUp(program, authority, buddyPDA, treasuryPDA) {
    var _a;
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var treasury, masterOrg, masterTreasuryPDA, masterTreasuryATA, buddy, profilePDA, profile, buddyTreasuryATA, referrerTreasuryATA;
        return tslib_1.__generator(this, function (_b) {
            switch (_b.label) {
                case 0: return [4 /*yield*/, program.account.treasury.fetchNullable(treasuryPDA)];
                case 1:
                    treasury = _b.sent();
                    if (!treasury)
                        throw type_1.TREASURY_DOESNT_EXIST;
                    if (!treasury.owners.find(function (owner) { return owner.ownerPda.toString() === buddyPDA.toString(); }))
                        throw type_1.TREASURY_OWNER_MISMATCH;
                    return [4 /*yield*/, program.account.masterOrganization.all()];
                case 2:
                    masterOrg = (_b.sent())[0];
                    masterTreasuryPDA = (0, pda_2.getMasterTreasuryPDA)(program, treasury.mint);
                    return [4 /*yield*/, (0, pda_1.getTokenAccount)(masterTreasuryPDA, treasury.mint)];
                case 3:
                    masterTreasuryATA = _b.sent();
                    return [4 /*yield*/, program.account.buddy.fetchNullable(buddyPDA)];
                case 4:
                    buddy = _b.sent();
                    if (!buddy)
                        throw type_1.BUDDY_DOESNT_EXIST;
                    profilePDA = buddyPDA;
                    if (!!((_a = buddy.buddyType) === null || _a === void 0 ? void 0 : _a.profile)) return [3 /*break*/, 6];
                    return [4 /*yield*/, program.account.buddy.all([
                            { memcmp: { offset: 8, bytes: buddy.authority.toBase58() } },
                        ])];
                case 5:
                    profile = (_b.sent())[0];
                    profilePDA = profile.publicKey;
                    _b.label = 6;
                case 6: return [4 /*yield*/, (0, pda_1.getTokenAccount)(treasuryPDA, treasury.mint)];
                case 7:
                    buddyTreasuryATA = _b.sent();
                    return [4 /*yield*/, (0, pda_1.getTokenAccount)(buddy.referrerTreasuryPda, treasury.mint)];
                case 8:
                    referrerTreasuryATA = _b.sent();
                    return [2 /*return*/, program.methods
                            .transferUp()
                            .accounts({
                            associatedTokenProgram: spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID,
                            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
                            systemProgram: anchor.web3.SystemProgram.programId,
                            rent: anchor.web3.SYSVAR_RENT_PUBKEY,
                            authority: authority,
                            masterOrganization: masterOrg.publicKey,
                            masterOrganizationTreasury: masterTreasuryPDA,
                            masterOrganizationTokenAccount: masterTreasuryATA,
                            buddyProfile: profilePDA,
                            buddy: buddyPDA,
                            buddyTreasury: treasuryPDA,
                            buddyTokenAccount: buddyTreasuryATA,
                            mint: (0, parse_1.parsePublicKey)(treasury.mint),
                            referrerTreasury: buddy.referrerTreasuryPda,
                            referrerTokenAccount: referrerTreasuryATA
                        })
                            .instruction()];
            }
        });
    });
}
exports.transferUp = transferUp;
